File: iid/gui/manager.py
#
# This file is part of the Infinite Improbability Drive.
#
# Copyright (C) 2009 by Jernej Kos <kostko@unimatrix-one.org>
# Copyright (C) 2009 by Anze Vavpetic <anze.vavpetic@gmail.com>
#
from OpenGL.GL import *
from OpenGL.GLU import *
import logging

# IID imports
from iid.events import EventDispatcher, EventType
from iid.gui.window import Titlebar, WindowFlags

# Logger for this module
logger = logging.getLogger(__name__)

class WindowManager(object):
  """
  Window manager is the top level manager for GUI events.
  """
  context = None
  windows = None
  widgetStyle = None
  
  # Current instance
  __manager = None
  
  def __init__(self, context):
    """
    Class constructor.
    
    @param context: A valid IID Context instance
    """
    WindowManager.__manager = self
    self.context = context
    self.context.scene.windowManager = self
    self.windows = []
    
    # Register signal handlers
    context.events.subscribe(EventType.MouseMove, self.event)
    context.events.subscribe(EventType.MousePress, self.event)
  
  @staticmethod
  def getManager():
    """
    Returns the current WindowManager instance.
    """
    return WindowManager.__manager
  
  def setWidgetStyle(self, style):
    """
    Sets a widget style to use.
    """
    self.widgetStyle = style
  
  def setFocus(self, window):
    """
    Sets the window that currently has focus by decreasing order
    to all windows previously above this window and setting this
    window's order to maximum value.
    """
    if len(self.windows):
      self.windows[0].active = False
      self.windows[0].emit("Window.lostFocus")
    
    try:
      self.windows.remove(window)
    except ValueError:
      pass
    
    self.windows.insert(0, window)
    window.active = True
    window.emit("Window.gotFocus")
  
  def registerWindow(self, window):
    """
    Registers a new window with this window manager.
    
    @param window: A valid Window instance
    """
    self.setFocus(window)
    
    if window.flags & WindowFlags.Titlebar:
      # Generate a titlebar
      window.titlebar = Titlebar(window)
  
  def event(self, event):
    """
    Event handler (dispatch events to widgets).
    """
    if event.eventType in (EventType.MouseMove, EventType.MousePress):
      # Dispatch to proper widget
      for window in self.windows:
        # Window titlebars are separate entities, since they are drawn
        # by the window manager and are not part of a window
        if window.titlebar.containsCoordinates(event.x, event.y):
          window.titlebar.event(event)
          break
        
        if window.containsCoordinates(event.x, event.y):
          window.event(event)
          break
      else:
        # No window accepted mouse press event, active window looses focus
        if len(self.windows) and event.eventType == EventType.MousePress:
          self.windows[0].active = False
          self.windows[0].emit("Window.lostFocus")
    elif event.eventType == EventType.Keyboard:
      # Get currently focused widget
      if len(self.windows) and self.windows[0].active:
        self.windows[0].event(event)
  
  def render(self):
    """
    Emits paint events to all visible top-level windows in reverse
    order (so windows below everything get painted first).
    """
    glEnable(GL_BLEND)
    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)
    
    glDisable(GL_DEPTH_TEST)
    glDisable(GL_LIGHTING)
    glDisable(GL_TEXTURE_2D)
    
    glMatrixMode(GL_PROJECTION)
    glPushMatrix()
    glLoadIdentity()
    gluOrtho2D(0, self.context.scene.width, 0, self.context.scene.height)
    glScalef(1, -1, 1)
    glTranslatef(0, -self.context.scene.height, 0)
    glMatrixMode(GL_MODELVIEW)
    glPushMatrix()
    glLoadIdentity()
    
    for window in self.windows[::-1]:
      if window.visible:
        self.__paintWindow(window)
    
    glPopMatrix()
    glMatrixMode(GL_PROJECTION)
    glPopMatrix()
    glMatrixMode(GL_MODELVIEW)
    
    glDisable(GL_BLEND)
    glEnable(GL_DEPTH_TEST)
    glEnable(GL_LIGHTING)
    glEnable(GL_TEXTURE_2D)
  
  def __paintWindow(self, window):
    """
    Paints a window to the screen.
    """
    if window.titlebar.visible:
      window.titlebar._paint()
    
    window._paint()
    

File: iid/src/gui/manager.cpp
/*
 * This file is part of the Infinite Improbability Drive.
 *
 * Copyright (C) 2009 by Jernej Kos <kostko@unimatrix-one.org>
 * Copyright (C) 2009 by Anze Vavpetic <anze.vavpetic@gmail.com>
 */
#include "gui/manager.h"
#include "gui/painter.h"
#include "gui/widget.h"
#include "gui/windows.h"
#include "gui/window.h"
#include "gui/button.h"
#include "events/mouse.h"
#include "events/keyboard.h"
#include "drivers/base.h"
#include "context.h"

namespace IID {

namespace GUI {

Manager *Manager::m_instance = 0;

Manager::Manager(Context *context)
  : m_context(context),
    m_root(new WindowManager()),
    m_mouseGrabWidget(m_root),
    m_keyboardGrabWidget(m_root)
{
  Vector2i dim = context->getViewportDimensions();
  m_instance = this;
  m_painter = new Painter(context->driver(), dim[0], dim[1]);
  m_root->setSize(dim[0] - 1, dim[1] - 1);
}

Manager::~Manager()
{
  delete m_painter;
  delete m_root;
}

void Manager::update(float dt)
{
}

void Manager::render()
{
  // Enter 2D mode
  m_context->driver()->enter2DMode();
  
  // Paint top-level widgets, events will propagate down the line
  m_painter->setClipping(true);
  m_painter->setClipRegion(0, 0, m_root->getWidth(), m_root->getHeight());
  m_root->paintEvent(m_painter);
  m_painter->setClipping(false);
  
  // Leave 2D mode
  m_context->driver()->leave2DMode();
}

void Manager::handleMouseMove(MouseEvent *event)
{
  m_mouseGrabWidget->event(event);
}

bool Manager::handleMousePress(MouseEvent *event)
{
  return m_mouseGrabWidget->event(event);
}

bool Manager::handleMouseRelease(MouseEvent *event)
{
  return m_mouseGrabWidget->event(event);
}

bool Manager::handleKeyboard(KeyboardEvent *event)
{
  return m_keyboardGrabWidget->event(event);
}

void Manager::grabMouse(Widget *widget)
{
  m_mouseGrabWidget = widget;
}

void Manager::ungrabMouse(Widget *widget)
{
  m_mouseGrabWidget = m_root;
}

void Manager::grabKeyboard(Widget *widget)
{
  m_keyboardGrabWidget = widget;
}

void Manager::ungrabKeyboard(Widget *widget)
{
  m_keyboardGrabWidget = m_root;
}

WindowManager *Manager::getWindowManager() const
{
  return m_root;
}

}

}
